### ISSUE TITLE
Variable lengths samples in batch in update()

### ISSUE THREAD

**Issue by Misterion777** [iffiX/machin#8]
Hey, i'm using DQN and my q-values are variable lengths sequences as I have different amount of actions for each state (my states have different shapes also). When sampling batches default Buffer concatenating them which leads to a tensor error. But when using update() with `concatenate_samples=False` it stills doesn't solve the problem as now samples are just lists, and all torch operations fail. Of course, I can pad the sequences, but then argmax() can return one of the padded indexes, as it's not possible to pass original lengths of each sample in batch in update() function. Is there any way to solve the problem right now, or it yet to be implemented?

**Comment by iffiX** [comment]
Hi, it would take a little bit more effort to adapt the framework to your case when ``concatenate_samples=False``, take ``DQNPer`` framework for example:

In the ``update`` function, it starts by sampling a list of batches:
```
       (batch_size,
         (state, action, reward, next_state, terminal, others),
         index, is_weight) = \
            self.replay_buffer.sample_batch(self.batch_size,
                                            concatenate_samples,
                                            sample_method="random_unique",
                                            sample_attrs=[
                                                "state", "action",
                                                "reward", "next_state",
                                                "terminal", "*"
                                            ])
```
As you have mentioned, things we are going to use are: ``state, action, reward, next_state, terminal``, which are ``Dict[str, list]`` for major attributes like state, action, next_state, ``list`` for sub-attributes like terminal and reward, since you have set  ``concatenate_samples=False``.

Below are the calls in the update function which utilizes these inputs:
```
next_q_value = self._criticize(next_state)
target_next_q_value = self._criticize(next_state, True)
action_value = q_value.gather(dim=1, index=self.action_get_function(action)
                              .to(device=q_value.device, dtype=t.long))
y_i = self.reward_function(
    reward, self.discount, target_next_q_value, terminal, others
)
```
1. ``_criticize`` call passes the state dictionary to your Q network, therefore your network should be able to deal with a list for each value in the state dict instead of one concatenated tensor. 
2. You need to overload the default ``self.action_get_function`` to make sure that it returns a long type tensor. The default implementation just extracts the value of key "action" from your action dictionary.
3. You also need to overload the default ``self.reward_function`` to make sure that reward is computed from your samples, remember the data type of each input mentioned above.

 Sorry for the inconvenience of this design, but since there are so many different models out there and it is hard to build one to satisfy all requirements, this is the best solution so far. 

If you have any better suggestions, or the above analysis doesn't work for you, please comment below :)

**Comment by Misterion777** [comment]
Thanks for the answer! Yes, I've been thinking to do what you are proposing but instead I came up with following solution: 
1. I've created custom buffer that is able to pad my sequences
```
class PaddingBuffer(Buffer):
    @staticmethod
    def make_tensor_from_batch(batch,device,concatenate):
        if concatenate and len(batch) != 0:
            item = batch[0]
            batch_size = len(batch)
            if t.is_tensor(item):
                batch = [it.to(device) for it in batch]             
                try:
                    batch = t.cat(batch, dim=0).to(device) # in case it's not possible to create batch
                except Exception:
                    batch = pad_sequence([it.squeeze(0) for it in batch],batch_first=True).to(device)
                return batch
            else:
                try:
                    return t.tensor(batch, device=device).view(batch_size, -1)
                except Exception:
                    raise ValueError("Batch not concatenable: {}"
                                     .format(batch))
        else:
            return batch
```
2. My network also receives initial lengths of the input. First it unpads input values according to lengths, then do all the computations and finally pads q-values again.

Both of these approaches are kinda weird, but maybe later there will be better idea how to do it.

**Comment by Misterion777** [comment]
Also, maybe I should open another issue, but I'll write here.
In docs for DQN `act_discrete_with_noise` shows implementation with epsilon, but when I'm running the code it seems it goes to another implementation with sampling from Categorical distribution. Which is fine, but as far as I understand there is no ability to control the randomness in this case as there is no epsilon. I propose to add some parameter _m_ (it can be same epsilon) to the sampling as such:
_Categorical(Softmax([Q(s, a)/m]))_

**Comment by iffiX** [comment]
Yeah, recently I pulled the original noise implementation in DQN series frameworks (DQNPer, RAINBOW, etc), and change it to the epsilon decay policy which is used by most studies:
https://github.com/iffiX/machin/blob/d0d6fafbeddab9dda8a9389a6be7ac5e98bc7762/machin/frame/algorithms/dqn.py#L252-L288
, the doc shows the latest implementation, but on pip distribution, the latest version is 0.3.4, I will skip version 0.3.5 and move to 0.4.0 since I am planning to release the PyTorch-lightning automatic launching utilities in that release, please be a little bit more patient so all things are properly tested and synced :)
